### PR TITLE
fix: fixed tests and logic to find the fastest endpoint

### LINKED ISSUE #32: Fix RPC tests
As seen in [this run](https://github.com/ubiquity/rpc-handler/actions/runs/10120429568/job/27990176188), the tests do not run successfully. There is also an issue with the run logging after the tests are done.
The tests should all be successful and the logs should not happen after Jest is torn down. This requires to investigate all the async calls and see what runs beyond the end of the tests.

### PR DESCRIPTION
**Pull request by gentlementlegen**
Resolves #32

### CODE REVIEW

**Review by Keyrxng — APPROVED** [review]
> The second issue is that an abort is called in the finally block which means all the requests get aborted regardless of their success.

My understanding is that `finally()` happens regardless of success or failure and so the results would be set and then the "open" `abortController` was closed during this block but I understand how `.race()` could return a failed call, nice catch @gentlementlegen. 

I too spent a lot of time trying to resolve this so hopefully it remains resolved this time

**Review by 0x4007 — COMMENTED** [review]
Code seems fine although I haven't used this codebase or ran the tests.

**Comment by gentlementlegen** [comment]
@Keyrxng @0x4007 I changed the logic a bit so instead of naively removing the first RPC of the list, the failing RPC is actually removed.

This also made me wonder, why do we test every endpoint when we are only interested in the fastest one, aka the first one to successfully fulfill the response? If we have to wait for slow RPCs, for example one that take 5 seconds to complete, doesn't it defeat the purpose of the function?

**Comment by 0x4007** [comment]
I agree we should just take the first working endpoint and terminate all the pending tests.

**Comment by gentlementlegen** [comment]
@0x4007 Maybe we can keep both functionalities if we ever need the full benchmarks but yeah I think it makes more sense. I'll take care of it in a separate task.

**Comment by 0x4007** [comment]
> if we ever need 

No plan for it, and theres no reason to add features "just in case"

**Comment by Keyrxng** [comment]
> This also made me wonder, why do we test every endpoint when we are only interested in the fastest one, aka the first one to successfully fulfill the response? If we have to wait for slow RPCs, for example one that take 5 seconds to complete, doesn't it defeat the purpose of the function?

If I understand correctly here - we will no longer gather a list of valid latencies/providers but instead will just use the first endpoint to successfully respond and use it? Does this imply then that the retry logic would be refactored such that we retest all RPCs again and not loop through our list of latencies?

I raised this same point previously which was why I introduced a timeout so that we kill any tests that run beyond the user defined limit but this way we still have a list of valid endpoints/latencies

**Comment by gentlementlegen** [comment]
I think there are different use cases where sometimes we need a list of endpoints (for example for retries), and sometimes just get the fastest one to respond.

For example: if you need the endpoint plenty of times but need to switch if it ends up failing, that would make sense to have a list of endpoints ordered by latency.

The use case I was solving for myself is that I need a valid RPC which is ideally fast but only once, so there is no need for me to evaluate all the endpoints as long as I get one that is decent and working.

**Comment by Keyrxng** [comment]
Ah that makes sense like a `getSingleFastestProvider()` approach

I looked at the latest jest [run](https://github.com/ubiquity/rpc-handler/actions/runs/10139976423/job/28034258991?pr=33) and the error has returned that you had me set out to resolve and I think was also the purpose of this PR?

> A worker process has failed to exit gracefully and has been force exited. This is likely caused by tests leaking due to improper teardown. Try running with --detectOpenHandles to find leaks. Active timers can also cause this, ensure that .unref() was called on them.

**Comment by gentlementlegen** [comment]
Seems the latest run that didn't have the issue was https://github.com/ubiquity/rpc-handler/actions/runs/10122809912
Will have a look.

**Comment by Keyrxng** [comment]
good luck, it's got hands

File: tests/benchmark.test.ts [hunk 1]
```diff
@@ -36,38 +35,34 @@ describe("RPCHandler", () => {
...
-      const rpcHandler = new module.RPCHandler({
-        ...testConfig,
-        rpcTimeout: 99999999,
-      });
-      await rpcHandler.testRpcPerformance();
-      const latencies = rpcHandler.getLatencies();
+    const module = await import("../types/rpc-handler");
+    const rpcHandler = new module.RPCHandler({
+      ...testConfig,
+      rpcTimeout: 99999999,
```
Comment (0x4007) [inline tests/benchmark.test.ts]: Seems like an unusual duration.

Comment (gentlementlegen) [reply tests/benchmark.test.ts]: Yes not sure why that was there, changed it to something that makes sense.

File: tests/call-handler.test.ts [hunk 2]
```diff
@@ -1,144 +1,18 @@
...
-
-        provider = await handler.getFastestRpcProvider();
-        const response = await provider.send("eth_call", [nonceBitmapData, "latest"]);
-        expect(response).toBeDefined();
-        expect(response).toBe("0x" + "00".repeat(32));
-      });
-    }, 15000);
-  });
-
-  describe("Write ops cases", () => {
```
Comment (Keyrxng) [inline tests/call-handler.test.ts:85]: Why have we removed all of the happy path tests, seems like we are only testing for failure cases?

**Comment by gentlementlegen** [comment]
@Keyrxng I solved the testing issue by mocking the `post` from `axios` because we do not need to hit the real endpoints anyway. Also I removed the tests where you were testing `JsonRpcProviders` because you are testing their API not ours, so I don't think that was really relevant. Everything is green now.

File: tests/call-handler.test.ts (hunk 2 above)
Comment (gentlementlegen) [reply tests/call-handler.test.ts:85]: Cf my comment on the main thread

File: tests/call-handler.test.ts [hunk 3]
```diff
@@ -186,47 +55,14 @@ describe("Call Handler", () => {
...
-      } catch (er) {
-        thrownError = er;
-      }
-
-      expect(thrownError).toBeInstanceOf(Error);
-      expect((thrownError as Error).message).toContain("invalid type: null");
-      expect(newHandler["_runtimeRpcs"].length).toBe(1);
-      expect(newHandler["_runtimeRpcs"][0]).toBe("http://127.0.0.1:8545");
+      const newProvider = await newHandler.getFastestRpcProvider();
+      expect(newProvider).toBeTruthy();
```
Comment (Keyrxng) [inline tests/call-handler.test.ts]: this test is no longer testing what it says it's testing

**Comment by Keyrxng** [comment]
> Also I removed the tests where you were testing JsonRpcProviders because you are testing their API not ours

I agree the first/top two tests were testing the provider API endpoints but only with `read ops` but the following `write ops` were all targeting the local anvil instance.

```ts
  const mods: HandlerConstructorConfig = {
      runtimeRpcs: rpcList.map((rpc) => rpc.url),
      networkRpcs: rpcList,
      autoStorage: false,
      cacheRefreshCycles: 1,
      networkName: "anvil",
      networkId: "31337",
      rpcTimeout: 700,
      proxySettings: {
        retryCount: 1,
        retryDelay: 10,
        logTier: "verbose",
        logger: new PrettyLogs(),
        strictLogs: false,
      },
    };
```

---

The provider racing was the original functionality of this package but it's now in control of all read and write operations so I think we should have tests for this

**Comment by gentlementlegen** [comment]
@Keyrxng I did my best to add back the tests with mods and naming them coherently, let me know what you think.

**Review by Keyrxng — CHANGES_REQUESTED** [review]
`rpc-handler.test.ts` contains tests related to the original functionality of the package, so things like proper init values and it returning the fastest provider.

`call-handler.test.ts` should contain tests that are testing the `proxy` and therefor all of the retry and reselection logic

As far as I can see all of the `call-handler.test` tests are now testing with `expect(provider).toBeTruthy();` which is only testing for the successful returning of a provider on init

Checkout this [jest run](https://github.com/ubiquity/rpc-handler/actions/runs/9981094118/job/27583941992) as I merged `proxify`. No leaking worker and lots of logs showing that the retry logic was being hit and was working. This [latest run](https://github.com/ubiquity/rpc-handler/actions/runs/10156775700/job/28085708779) only shows `provider initialized` logs and doesn't appear to be hitting the retry logic

**Comment by gentlementlegen** [comment]
@Keyrxng Right ok I get it. But then I should try to entirely mock the RPC provider logic because we do not want real endpoints to be hit.

**Comment by Keyrxng** [comment]
> @Keyrxng Right ok I get it. But then I should try to entirely mock the RPC provider logic because we do not want real endpoints to be hit.

Well no actually because anvil is dumb and will respond successfully more or less no matter what you send to it, so in the specific case that our nonceBitMap read op was failing and it was only for certain providers anvil will always give you a great light under that scenario

Maybe I am mistaken @gentlementlegen and you can do that it may be possible so let me correct above and just say that the above is my opinion and that's why I took the approach that I did but you can mock everything if we are still able to effectively test the call-handler aspect I guess

**Comment by gentlementlegen** [comment]
@Keyrxng Took another different approach: I removed the mocks and always use Anvil instead, tell me what you think.

**Review by Keyrxng — APPROVED** [review]
Everything looks good as far as I can tell. Test logs showing retry logic and no leaky worker error

**Pull request merged**